### PR TITLE
cloud: add retry on web identity credentials

### PR DESCRIPTION
**Pull request by 3pointer**
Signed-off-by: 3pointer <luancheng@pingcap.com>

### What is changed and how it works?

Issue Number: Close https://github.com/tikv/tikv/issues/13122

What's Changed:
1. In EKS environment. we shouldn't let the credential fail back. if we detected the credential provider from k8s env. and if we gets error. just retry it.
2. Add retry time to 14(about 5 mins) for all s3 interface and credentials interface. according to experience the jitter of network could be 1 min

```cloud: retry on web identity provider credentials
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test
1. create eks cluster.
2. import data.
3. make sure SA with correct role.
origin_img_v2_51eda314-8a34-4b25-837c-81fb009f00bg
4. do backup check the backup succeed
origin_img_v2_2987666a-5009-44e9-948f-201729a5b4cg
5. bind SA with invalid role
origin_img_v2_aa63a7e8-28aa-47f6-8b6d-7ee275e0516g
6. do backup and check the error message
origin_img_v2_234ac260-dce1-4d6e-a90f-266e762ac20g

it doesn't fall back to default credentials provider.

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Fix the issue that cause permission denied when gets error from web identity provider and fail back to default provider.
```

### CODE REVIEW

File: components/cloud/aws/src/util.rs [hunk 1]
```diff
@@ -92,21 +95,29 @@ impl Default for DefaultCredentialsProvider {
...
+                        break Err(CredentialsError::new(format_args!(
+                            "Couldn't find AWS credentials in sources ({}).",
+                            e.message
+                        )));
+                    }
+                }
+            };
+            retry_cnt += 1;
+            // this could happen in web identity provider, we need retry it.
+            thread::sleep(time::Duration::from_secs(retry_cnt));
```
Comment (YuJuncen) [inline components/cloud/aws/src/util.rs:119]: Please use `tokio::time::sleep`, or we may block the runtime.
```suggestion
            tokio::time::sleep(time::Duration::from_secs(retry_cnt)).await;
```

**Review by YuJuncen — APPROVED** [review]
rest lgtm

File: components/cloud/aws/src/util.rs [hunk 2]
```diff
@@ -92,21 +110,69 @@ impl Default for DefaultCredentialsProvider {
...
+    #[cfg(feature = "failpoints")]
+    #[tokio::test]
+    async fn test_default_provider() {
+        let default_provider = DefaultCredentialsProvider::default();
+        std::env::set_var(AWS_WEB_IDENTITY_TOKEN_FILE, "tmp");
+        // mock k8s env with web_identitiy_provider
+        fail::cfg("cred_err", "return").unwrap();
+        fail::cfg("retry_count", "return(1)").unwrap();
+        let res = default_provider.credentials().await;
+        assert_eq!(res.is_err(), true);
```
Comment (YuJuncen) [inline components/cloud/aws/src/util.rs:168]:
```suggestion
        assert!(res.is_err());
```

**Review by YuJuncen — APPROVED** [review]

File: components/cloud/aws/src/util.rs [hunk 3]
```diff
@@ -49,6 +77,22 @@ pub fn get_region(region: &str, endpoint: &str) -> io::Result<Region> {
...
+    G: FnMut() -> F,
+    F: Future<Output = Result<T, E>>,
+    E: RetryError + std::fmt::Display,
+{
+    let id = uuid::Uuid::new_v4();
+    retry_ext(
+        action,
+        RetryExt::default().with_fail_hook(move |err: &E| {
+            warn!("aws request meet error."; "err" => %err, "retry?" => %err.is_retryable(), "context" => %name, "uuid" => %id);
+            metrics::CLOUD_ERROR_VEC.with_label_values(&["aws", name]);
```
Comment (YuJuncen) [inline components/cloud/aws/src/util.rs:91]:
```suggestion
            metrics::CLOUD_ERROR_VEC.with_label_values(&["aws", name]).inc();
```

**Review by hicqu — APPROVED** [review]

**Comment by hicqu** [comment]
@3pointer please fix DCO check.

**Pull request merged**